### ISSUE TITLE
Security: support Modular Crypt Format or its successor PHC

### ISSUE THREAD

**Issue by FroMage** [quarkusio/quarkus#5667]
ATM our security extensions support several sorts of algorithms:

- security-properties says clear or `HEX( MD5( username ":" realm ":" password ) )`
- security-jdbc says clear or bCrypt (but not in MCF/PHC format: all fields separated)

We should support [MCF](https://passlib.readthedocs.io/en/stable/modular_crypt_format.html) and/or its successor [PHC](https://github.com/P-H-C/phc-string-format/blob/master/phc-sf-spec.md) for both options, even if we don't support all the algo types.

**Comment by FroMage** [comment]
https://www.npmjs.com/package/@phc/bcrypt has examples of the PHC format for bcrypt.

**Comment by FroMage** [comment]
http://wildfly-security.github.io/wildfly-elytron/1.9.x/api-javadoc/org/wildfly/security/password/util/ModularCrypt.html has support for MCF, so it's just a matter of being able to use it.

As for PHC, we probably need to add support for it in Elytron, WDYT @dmlloyd ?

**Comment by dmlloyd** [comment]
If PHC is, as they claim, a "well defined subset of the Modular Crypt Format" then it's just a question of adding support for whatever additional formats we are missing.  The PHC document at https://github.com/P-H-C/phc-string-format/blob/master/phc-sf-spec.md mentions a few password types which are not supported presently by Elytron, specifically the "argon" group of algorithms.

/cc @darranl, @fjuma

**Comment by sberyozkin** [comment]
Hey @FroMage, can you please discuss with @wowselim how to proceed here, he is interested in this enhancement, see #38634, thanks.

**Comment by wowselim** [comment]
@FroMage you're right, it's supported indeed. Here's a project with a working test to demonstrate:

https://github.com/wowselim/quarkus-jdbc-auth-reproducer

**Comment by FroMage** [comment]
Awesome!

So the config required is:

```properties
quarkus.security.jdbc.enabled=true
quarkus.security.jdbc.principal-query.sql=SELECT password, role FROM account WHERE email=?
quarkus.security.jdbc.principal-query.bcrypt-password-mapper.enabled=true
quarkus.security.jdbc.principal-query.bcrypt-password-mapper.password-index=1
quarkus.security.jdbc.principal-query.bcrypt-password-mapper.salt-index=-1
quarkus.security.jdbc.principal-query.bcrypt-password-mapper.iteration-count-index=-1
```

I think we should document this, and also make `salt-index` and `iteration-count-index` non-required and default to `-1` like they do upstream.

WDYT @sberyozkin ?

**Comment by dmlloyd** [comment]
It's probably best to change their type to `OptionalInt` then, and only accept `-1` on a compatibility basis.

**Comment by FroMage** [comment]
Sure.

**Comment by FroMage** [comment]
The docs are great, but we should also fix the default values of the salt/iteration config values to simplify.
